python-package/xgboost/core.py

```diff
@@ -138,27 +138,42 @@ def c_array(ctype, values):
     return (ctype * len(values))(*values)
 
 
-def _maybe_from_pandas(data, feature_names, feature_types):
+def _maybe_from_pandas(data, label, feature_names, feature_types):
     """ Extract internal data from pd.DataFrame """
     try:
         import pandas as pd
     except ImportError:
-        return data, feature_names, feature_types
+        return data, label, feature_names, feature_types
 
     if not isinstance(data, pd.DataFrame):
-        return data, feature_names, feature_types
-
-    dtypes = data.dtypes
-    if not all(dtype.name in ('int64', 'float64', 'bool') for dtype in dtypes):
-        raise ValueError('DataFrame.dtypes must be int, float or bool')
+        return data, label, feature_names, feature_types
+
+    mapper = {'int8': 'int', 'int16': 'int', 'int32': 'int', 'int64': 'int',
+              'uint8': 'int', 'uint16': 'int', 'uint32': 'int', 'uint64': 'int',
+              'float16': 'float', 'float32': 'float', 'float64': 'float',
+              'bool': 'i'}
+
+    data_dtypes = data.dtypes
+    if not all(dtype.name in (mapper.keys()) for dtype in data_dtypes):
+        raise ValueError('DataFrame.dtypes for data must be int, float or bool')
+
+    if label is not None:
+        if isinstance(label, pd.DataFrame):
```

Comment:
@terrytangyuan The DMatrix class in current master checks the ```data``` variable. If the ```data``` is a ```DataFrame```, the ```data```, ```feature_names``` and ```feature_types``` get checked by the ```_maybe_from_pandas``` function. The ```label``` gets processed in the DMatrix class and will always generate an error if the ```label``` is a named ```DataFrame```.

I think people who use XGBoost and Pandas, usually generate two ```DataFrames```, one for X and one for y. They should not have a problem passing both to a DMatrix. That is why after merge, the ```label``` gets an extra check, but only if the ```data``` variable is a ```DataFrame```. The example mentioned will not work, because the ```data``` is a ```numpy.ndarray``` and the ```label``` is a Pandas ```DataFrame```, so it will not get passed through to the ```_maybe_from_pandas``` function.

@sinhrks I think you made the ```_maybe_from_pandas``` function to check various variables if the ```data``` variable is ```DataFrame```. We can re-code the DMatrix class and let it check the ```data```, ```label```, ```feature_names``` and ```feature_types``` separately if you want... I just don't think that is necessary. It is your call. 

```
data            - label

numpy.ndarray   - numpy.ndarray   - works in current master
named DataFrame - numpy.ndarray   - works in current master
named DataFrame - named DataFrame - only works after merge
numpy.ndarray   - named DataFrame - does not work
```

Guys, I just want to help and it feels like I am stepping on some toes... please don't see it like that.